File: routes/network_control.py
from flask import Blueprint, request, jsonify
from utils.auth import token_required, admin_required
from models.voucher import Voucher
from models.router import Router
from models.network import Network
from utils.router_manager import get_router_manager
from database import db
import json
from datetime import datetime, timedelta

network_control_bp = Blueprint('network_control', __name__)

@network_control_bp.route('/routers', methods=['GET'])
@token_required
def get_routers(current_user):
    """Get all configured routers"""
    try:
        routers = Router.query.all()
        return jsonify([router.to_dict() for router in routers])
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@network_control_bp.route('/routers', methods=['POST'])
@admin_required
def add_router(current_user):
    """Add new router configuration"""
    try:
        data = request.get_json()
        
        router = Router()
        router.name = data.get('name')
        router.brand = data.get('brand')
        router.model = data.get('model')
        router.ip_address = data.get('ip_address')
        router.username = data.get('username')
        router.password = data.get('password')
        router.api_port = data.get('api_port')
        router.is_active = True
        
        # Test connection
        try:
            manager = get_router_manager(router)
            if manager.test_connection():
                router.status = 'connected'
                router.last_seen = datetime.utcnow()
            else:
                router.status = 'disconnected'
        except Exception:
            router.status = 'error'
        
        db.session.add(router)
        db.session.commit()
        
        return jsonify({
            'message': 'تم إضافة الراوتر بنجاح',
            'router': router.to_dict()
        })
        
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@network_control_bp.route('/routers/<int:router_id>/test', methods=['POST'])
@admin_required
def test_router_connection(current_user, router_id):
    """Test router connection"""
    try:
        router = Router.query.get_or_404(router_id)
        
        manager = get_router_manager(router)
        connected = manager.test_connection()
        
        if connected:
            router.status = 'connected'
            router.last_seen = datetime.utcnow()
            message = 'الاتصال بالراوتر ناجح'
        else:
            router.status = 'disconnected'
            message = 'فشل في الاتصال بالراوتر'
        
        db.session.commit()
        
        return jsonify({
            'connected': connected,
            'message': message,
            'router': router.to_dict()
        })
        
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@network_control_bp.route('/vouchers/create_batch', methods=['POST'])
@admin_required
def create_voucher_batch(current_user):
    """Create batch of vouchers with advanced options"""
    try:
        data = request.get_json()
        
        quantity = int(data.get('quantity', 1))
        batch_id = f"BATCH_{datetime.utcnow().strftime('%Y%m%d_%H%M%S')}"
        
        # Voucher configuration
        duration_hours = int(data.get('duration_hours', 24))
        data_limit_mb = data.get('data_limit_mb')
        speed_limit_kbps = data.get('speed_limit_kbps')
        voucher_type = data.get('voucher_type', 'standard')
        price = float(data.get('price', 0))
        
        vouchers = []
        
        for i in range(quantity):
            voucher = Voucher()
            voucher.batch_id = batch_id
            voucher.duration_hours = duration_hours
            voucher.data_limit_mb = int(data_limit_mb) if data_limit_mb else None
            voucher.speed_limit_kbps = int(speed_limit_kbps) if speed_limit_kbps else None
            voucher.voucher_type = voucher_type
            voucher.price = price
            voucher.created_by = current_user.id
            
            # Set expiration date for voucher usage
            if data.get('voucher_expires_days'):
                voucher.expires_at = datetime.utcnow() + timedelta(days=int(data.get('voucher_expires_days')))
            
            # Network restrictions
            if data.get('allowed_networks'):
                voucher.allowed_networks = json.dumps(data.get('allowed_networks'))
            
            # Generate QR code
            base_url = data.get('base_url', 'http://localhost:5000')
            voucher.generate_qr_data(base_url)
            
            db.session.add(voucher)
            vouchers.append(voucher)
        
        db.session.commit()
        
        return jsonify({
            'message': f'تم إنشاء {quantity} كارت بنجاح',
            'batch_id': batch_id,
            'vouchers': [v.to_dict() for v in vouchers]
        })
        
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@network_control_bp.route('/vouchers/<voucher_code>/activate', methods=['POST'])
def activate_voucher(voucher_code):
    """Activate voucher and grant network access"""
    try:
        data = request.get_json()
        client_mac = data.get('client_mac')
        client_ip = data.get('client_ip')
        
        voucher = Voucher.query.filter_by(code=voucher_code).first()
        
        if not voucher:
            return jsonify({'error': 'كود الكارت غير صحيح'}), 404
        
        if not voucher.is_valid():
            return jsonify({'error': 'كارت الاتصال منتهي الصلاحية أو مستخدم'}), 400
        
        # Mark voucher as used
        voucher.mark_as_used()
        voucher.session_start = datetime.utcnow()
        voucher.client_mac = client_mac
        voucher.client_ip = client_ip
        
        # Calculate session end time
        if voucher.duration_hours:
            voucher.session_end = datetime.utcnow() + timedelta(hours=voucher.duration_hours)
        
        # Apply voucher to router(s)
        routers = Router.query.filter_by(is_active=True).all()
        
        for router in routers:
            try:
                manager = get_router_manager(router)
                if manager.connect():
                    # Add user to router's hotspot/guest system
                    if router.brand == 'MikroTik':
                        success = manager.add_hotspot_user(
                            username=voucher_code,
                            password=voucher.session_token,
                            profile='voucher_profile'
                        )
                    elif router.brand == 'Ubiquiti':
                        success = manager.add_guest_user(
                            username=voucher_code,
                            password=voucher.session_token,
                            duration_minutes=voucher.duration_hours * 60 if voucher.duration_hours else 1440
                        )
                    else:
                        success = False
                    
                    if success:
                        router.last_seen = datetime.utcnow()
                        router.status = 'connected'
                
                manager.disconnect()
                
            except Exception as e:
                print(f"Error applying voucher to router {router.name}: {e}")
        
        db.session.commit()
        
        return jsonify({
            'message': 'تم تفعيل كارت الاتصال بنجاح',
            'session_token': voucher.session_token,
            'duration_hours': voucher.duration_hours,
            'data_limit_mb': voucher.data_limit_mb,
            'speed_limit_kbps': voucher.speed_limit_kbps,
            'session_end': voucher.session_end.isoformat() if voucher.session_end else None
        })
        
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@network_control_bp.route('/vouchers/<voucher_code>/usage', methods=['GET'])
def get_voucher_usage(voucher_code):
    """Get voucher usage statistics"""
    try:
        voucher = Voucher.query.filter_by(code=voucher_code).first()
        
        if not voucher:
            return jsonify({'error': 'كود الكارت غير صحيح'}), 404
        
        usage_info = {
            'code': voucher.code,
            'status': voucher.status,
            'data_used_mb': voucher.data_used_mb,
            'data_limit_mb': voucher.data_limit_mb,
            'session_start': voucher.session_start.isoformat() if voucher.session_start else None,
            'session_end': voucher.session_end.isoformat() if voucher.session_end else None,
            'remaining_time': None,
            'remaining_data': None
        }
        
        # Calculate remaining time
        if voucher.session_end and voucher.status == 'used':
            remaining_seconds = (voucher.session_end - datetime.utcnow()).total_seconds()
            if remaining_seconds > 0:
                usage_info['remaining_time'] = int(remaining_seconds / 60)  # in minutes
            else:
                usage_info['remaining_time'] = 0
        
        # Calculate remaining data
        if voucher.data_limit_mb:
            usage_info['remaining_data'] = max(0, voucher.data_limit_mb - voucher.data_used_mb)
        
        return jsonify(usage_info)
        
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@network_control_bp.route('/vouchers/<voucher_code>/disconnect', methods=['POST'])
@admin_required
def disconnect_voucher(current_user, voucher_code):
    """Disconnect voucher session"""
    try:
        voucher = Voucher.query.filter_by(code=voucher_code).first()
        
        if not voucher:
            return jsonify({'error': 'كود الكارت غير صحيح'}), 404
        
        if voucher.status != 'used':
            return jsonify({'error': 'الكارت غير نشط'}), 400
        
        # Remove from routers
        routers = Router.query.filter_by(is_active=True).all()
        
        for router in routers:
            try:
                manager = get_router_manager(router)
                if manager.connect():
                    if router.brand == 'MikroTik':
                        manager.remove_hotspot_user(voucher_code)
                    # Add disconnect logic for other router types
                
                manager.disconnect()
                
            except Exception as e:
                print(f"Error disconnecting voucher from router {router.name}: {e}")
        
        # Mark voucher as expired
        voucher.status = 'expired'
        voucher.session_end = datetime.utcnow()
        
        db.session.commit()
        
        return jsonify({'message': 'تم قطع الاتصال بنجاح'})
        
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@network_control_bp.route('/network/clients', methods=['GET'])
@token_required
def get_connected_clients(current_user):
    """Get list of connected clients"""
    try:
        # Get active vouchers (connected clients)
        active_vouchers = Voucher.query.filter_by(status='used').filter(
            Voucher.session_end > datetime.utcnow()
        ).all()
        
        clients = []
        for voucher in active_vouchers:
            if voucher.session_start and voucher.session_end:
                remaining_time = (voucher.session_end - datetime.utcnow()).total_seconds() / 60
                if remaining_time > 0:
                    clients.append({
                        'voucher_code': voucher.code,
                        'client_mac': voucher.client_mac,
                        'client_ip': voucher.client_ip,
                        'session_start': voucher.session_start.isoformat(),
                        'remaining_minutes': int(remaining_time),
                        'data_used_mb': voucher.data_used_mb,
                        'data_limit_mb': voucher.data_limit_mb
                    })
        
        return jsonify({
            'connected_clients': len(clients),
            'clients': clients
        })
        
    except Exception as e:
        return jsonify({'error': str(e)}), 500

File: static/js/network_control.js
// Network Control JavaScript

let routers = [];
let clients = [];
let networkStats = {};

// Load routers
async function loadRouters() {
    try {
        showLoading();
        const response = await fetch('/api/control/routers', {
            headers: {
                'Authorization': 'Bearer ' + getToken()
            }
        });
        
        if (response.ok) {
            routers = await response.json();
            displayRouters();
            updateRouterStats();
        } else {
            showError('فشل في تحميل بيانات الراوترات');
        }
    } catch (error) {
        showError('حدث خطأ في تحميل الراوترات: ' + error.message);
    } finally {
        hideLoading();
    }
}

// Display routers in table
function displayRouters() {
    const tbody = document.getElementById('routers-tbody');
    tbody.innerHTML = '';
    
    routers.forEach(router => {
        const row = document.createElement('tr');
        
        const statusClass = router.status === 'connected' ? 'status-success' : 
                          router.status === 'disconnected' ? 'status-warning' : 'status-error';
        
        const statusText = router.status === 'connected' ? 'متصل' : 
                          router.status === 'disconnected' ? 'غير متصل' : 'خطأ';
        
        const lastSeen = router.last_seen ? 
            new Date(router.last_seen).toLocaleString('ar-SA') : 'غير متاح';
        
        row.innerHTML = `
            <td>${router.name}</td>
            <td>${router.brand} ${router.model || ''}</td>
            <td>${router.ip_address}</td>
            <td><span class="status-badge ${statusClass}">${statusText}</span></td>
            <td>${lastSeen}</td>
            <td>
                <div class="action-buttons">
                    <button class="btn btn-sm btn-info" onclick="testRouterConnection(${router.id})" title="اختبار الاتصال">
                        <i data-feather="wifi"></i>
                    </button>
                    <button class="btn btn-sm btn-secondary" onclick="editRouter(${router.id})" title="تعديل">
                        <i data-feather="edit"></i>
                    </button>
                    <button class="btn btn-sm btn-danger" onclick="deleteRouter(${router.id})" title="حذف">
                        <i data-feather="trash-2"></i>
                    </button>
                </div>
            </td>
        `;
        tbody.appendChild(row);
    });
    
    feather.replace();
}

// Load connected clients
async function loadClients() {
    try {
        const response = await fetch('/api/control/network/clients', {
            headers: {
                'Authorization': 'Bearer ' + getToken()
            }
        });
        
        if (response.ok) {
            const data = await response.json();
            clients = data.clients;
            displayClients();
            updateClientStats(data.connected_clients);
        } else {
            showError('فشل في تحميل بيانات المستخدمين');
        }
    } catch (error) {
        showError('حدث خطأ في تحميل بيانات المستخدمين: ' + error.message);
    }
}

// Display clients in table
function displayClients() {
    const tbody = document.getElementById('clients-tbody');
    tbody.innerHTML = '';
    
    clients.forEach(client => {
        const row = document.createElement('tr');
        
        const remainingTime = client.remaining_minutes > 60 ? 
            Math.floor(client.remaining_minutes / 60) + ' ساعة ' + (client.remaining_minutes % 60) + ' دقيقة' :
            client.remaining_minutes + ' دقيقة';
        
        const dataUsed = client.data_used_mb ? 
            client.data_used_mb.toFixed(2) + ' MB' + 
            (client.data_limit_mb ? ' / ' + client.data_limit_mb + ' MB' : ' / غير محدود') :
            '0 MB';
        
        row.innerHTML = `
            <td>${client.voucher_code}</td>
            <td>${client.client_mac || 'غير متاح'}</td>
            <td>${client.client_ip || 'غير متاح'}</td>
            <td>${new Date(client.session_start).toLocaleString('ar-SA')}</td>
            <td>${remainingTime}</td>
            <td>${dataUsed}</td>
            <td>
                <div class="action-buttons">
                    <button class="btn btn-sm btn-info" onclick="viewClientDetails('${client.voucher_code}')" title="التفاصيل">
                        <i data-feather="info"></i>
                    </button>
                    <button class="btn btn-sm btn-danger" onclick="disconnectClient('${client.voucher_code}')" title="قطع الاتصال">
                        <i data-feather="x-circle"></i>
                    </button>
                </div>
            </td>
        `;
        tbody.appendChild(row);
    });
    
    feather.replace();
}

// Test router connection
async function testRouterConnection(routerId) {
    try {
        showLoading();
        const response = await fetch(`/api/control/routers/${routerId}/test`, {
            method: 'POST',
            headers: {
                'Authorization': 'Bearer ' + getToken()
            }
        });
        
        const data = await response.json();
        
        if (response.ok) {
            if (data.connected) {
                showSuccess(data.message);
            } else {
                showWarning(data.message);
            }
            loadRouters(); // Refresh router list
        } else {
            showError(data.error || 'فشل في اختبار الاتصال');
        }
    } catch (error) {
        showError('حدث خطأ في اختبار الاتصال: ' + error.message);
    } finally {
        hideLoading();
    }
}

// Add new router
async function addRouter() {
    const name = document.getElementById('router-name').value;
    const brand = document.getElementById('router-brand').value;
    const model = document.getElementById('router-model').value;
    const ip_address = document.getElementById('router-ip').value;
    const username = document.getElementById('router-username').value;
    const password = document.getElementById('router-password').value;
    const api_port = document.getElementById('router-port').value;
    
    if (!name || !brand || !ip_address || !username || !password) {
        showError('يرجى ملء جميع الحقول المطلوبة');
        return;
    }
    
    try {
        showLoading();
        const response = await fetch('/api/control/routers', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': 'Bearer ' + getToken()
            },
            body: JSON.stringify({
                name,
                brand,
                model,
                ip_address,
                username,
                password,
                api_port: api_port ? parseInt(api_port) : null
            })
        });
        
        const data = await response.json();
        
        if (response.ok) {
            showSuccess(data.message);
            hideAddRouterModal();
            loadRouters();
            document.getElementById('add-router-form').reset();
        } else {
            showError(data.error || 'فشل في إضافة الراوتر');
        }
    } catch (error) {
        showError('حدث خطأ في إضافة الراوتر: ' + error.message);
    } finally {
        hideLoading();
    }
}

// Create voucher batch with advanced options
async function createVoucherBatch() {
    const quantity = parseInt(document.getElementById('batch-quantity').value);
    const duration_hours = parseInt(document.getElementById('batch-duration').value);
    const data_limit_mb = document.getElementById('batch-data-limit').value;
    const speed_limit_kbps = document.getElementById('batch-speed-limit').value;
    const voucher_type = document.getElementById('batch-voucher-type').value;
    const price = parseFloat(document.getElementById('batch-price').value);
    const voucher_expires_days = parseInt(document.getElementById('batch-expires-days').value);
    
    if (!quantity || !duration_hours || quantity <= 0 || quantity > 1000) {
        showError('يرجى إدخال عدد صحيح للكروت (1-1000)');
        return;
    }
    
    try {
        showLoading();
        const requestData = {
            quantity,
            duration_hours,
            voucher_type,
            price,
            voucher_expires_days,
            base_url: window.location.origin
        };
        
        if (data_limit_mb) requestData.data_limit_mb = parseInt(data_limit_mb);
        if (speed_limit_kbps) requestData.speed_limit_kbps = parseInt(speed_limit_kbps);
        
        const response = await fetch('/api/control/vouchers/create_batch', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': 'Bearer ' + getToken()
            },
            body: JSON.stringify(requestData)
        });
        
        const data = await response.json();
        
        if (response.ok) {
            showSuccess(data.message);
            
            // Show batch details
            const batchInfo = `
                <div class="batch-info">
                    <h4>تفاصيل الدفعة</h4>
                    <p><strong>رقم الدفعة:</strong> ${data.batch_id}</p>
                    <p><strong>عدد الكروت:</strong> ${quantity}</p>
                    <p><strong>مدة كل كارت:</strong> ${duration_hours} ساعة</p>
                    ${data_limit_mb ? `<p><strong>حد البيانات:</strong> ${data_limit_mb} ميجابايت</p>` : ''}
                    ${speed_limit_kbps ? `<p><strong>حد السرعة:</strong> ${speed_limit_kbps} كيلوبايت/ثانية</p>` : ''}
                </div>
            `;
            
            showInfo('تم إنشاء الكروت بنجاح!<br>' + batchInfo);
            
            // Reset form
            document.getElementById('batch-quantity').value = 10;
            document.getElementById('batch-duration').value = 24;
            document.getElementById('batch-data-limit').value = '';
            document.getElementById('batch-speed-limit').value = '';
            document.getElementById('batch-voucher-type').value = 'standard';
            document.getElementById('batch-price').value = 0;
            
        } else {
            showError(data.error || 'فشل في إنشاء الكروت');
        }
    } catch (error) {
        showError('حدث خطأ في إنشاء الكروت: ' + error.message);
    } finally {
        hideLoading();
    }
}

// Disconnect client
async function disconnectClient(voucherCode) {
    if (!confirm('هل تريد قطع اتصال هذا المستخدم؟')) return;
    
    try {
        showLoading();
        const response = await fetch(`/api/control/vouchers/${voucherCode}/disconnect`, {
            method: 'POST',
            headers: {
                'Authorization': 'Bearer ' + getToken()
            }
        });
        
        const data = await response.json();
        
        if (response.ok) {
            showSuccess(data.message);
            loadClients(); // Refresh clients list
        } else {
            showError(data.error || 'فشل في قطع الاتصال');
        }
    } catch (error) {
        showError('حدث خطأ في قطع الاتصال: ' + error.message);
    } finally {
        hideLoading();
    }
}

// View client details
async function viewClientDetails(voucherCode) {
    try {
        const response = await fetch(`/api/control/vouchers/${voucherCode}/usage`, {
            headers: {
                'Authorization': 'Bearer ' + getToken()
            }
        });
        
        if (response.ok) {
            const data = await response.json();
            
            const details = `
                <div class="client-details">
                    <h4>تفاصيل الاتصال</h4>
                    <p><strong>كود الكارت:</strong> ${data.code}</p>
                    <p><strong>الحالة:</strong> ${data.status === 'used' ? 'نشط' : data.status}</p>
                    <p><strong>البيانات المستخدمة:</strong> ${data.data_used_mb.toFixed(2)} MB</p>
                    ${data.data_limit_mb ? `<p><strong>حد البيانات:</strong> ${data.data_limit_mb} MB</p>` : ''}
                    ${data.remaining_data !== null ? `<p><strong>البيانات المتبقية:</strong> ${data.remaining_data.toFixed(2)} MB</p>` : ''}
                    ${data.remaining_time !== null ? `<p><strong>الوقت المتبقي:</strong> ${data.remaining_time} دقيقة</p>` : ''}
                    ${data.session_start ? `<p><strong>بداية الجلسة:</strong> ${new Date(data.session_start).toLocaleString('ar-SA')}</p>` : ''}
                    ${data.session_end ? `<p><strong>نهاية الجلسة:</strong> ${new Date(data.session_end).toLocaleString('ar-SA')}</p>` : ''}
                </div>
            `;
            
            showInfo(details);
        } else {
            showError('فشل في تحميل تفاصيل المستخدم');
        }
    } catch (error) {
        showError('حدث خطأ في تحميل التفاصيل: ' + error.message);
    }
}

// Load network statistics
async function loadNetworkStats() {
    try {
        const response = await fetch('/api/stats/dashboard', {
            headers: {
                'Authorization': 'Bearer ' + getToken()
            }
        });
        
        if (response.ok) {
            networkStats = await response.json();
            updateNetworkStats();
        }
    } catch (error) {
        console.error('Error loading network stats:', error);
    }
}

// Update statistics display
function updateNetworkStats() {
    document.getElementById('active-sessions').textContent = 
        clients.filter(c => c.remaining_minutes > 0).length;
    
    const totalDataUsed = clients.reduce((sum, client) => sum + (client.data_used_mb || 0), 0);
    document.getElementById('data-usage').textContent = totalDataUsed.toFixed(2) + ' MB';
}

function updateRouterStats() {
    const connectedRouters = routers.filter(r => r.status === 'connected').length;
    document.getElementById('connected-routers').textContent = connectedRouters;
}

function updateClientStats(connectedCount) {
    document.getElementById('active-clients').textContent = connectedCount;
}

// Refresh functions
function refreshAll() {
    loadRouters();
    loadClients();
    loadNetworkStats();
}

function refreshClients() {
    loadClients();
}

// Modal functions
function showAddRouterModal() {
    document.getElementById('add-router-modal').style.display = 'block';
}

function hideAddRouterModal() {
    document.getElementById('add-router-modal').style.display = 'none';
}

// Auto-refresh every 30 seconds
setInterval(() => {
    loadClients();
    loadNetworkStats();
}, 30000);

// Auto-refresh router status every 2 minutes
setInterval(() => {
    loadRouters();
}, 120000);